File: app/src/main/java/com/example/moviesearcher/model/repositories/CategoryListRepository.java
package com.example.moviesearcher.model.repositories;

import com.example.moviesearcher.model.data.Category;
import com.example.moviesearcher.model.data.Subcategory;
import com.example.moviesearcher.model.services.MovieDbApiService;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

public class CategoryListRepository {

    private List<Category> categoryList = new ArrayList<>();

    public interface CategoryListResponse{
        void onProcessFinished(List<Category> categoryList);
    }

    public void fetchData(CategoryListResponse callback){
        new MovieDbApiService().getGenres(genresMap -> {

            categoryList.add(genreMapToCategory(genresMap));

            new MovieDbApiService().getLanguages(subcategoryList -> {
                sortList(subcategoryList);
                categoryList.add(new Category("Language", subcategoryList));
                callback.onProcessFinished(categoryList);
            });
        });
    }

    private Category genreMapToCategory(HashMap<Integer, String> genresMap){
        List<Integer> genreIds = new ArrayList<>(genresMap.keySet());
        List<String> genreNames = new ArrayList<>(genresMap.values());

        List<Subcategory> subcategoryList = new ArrayList<>();
        for (int i = 0; i < genreIds.size(); i++){
            subcategoryList.add(new Subcategory(genreIds.get(i), genreNames.get(i)));
        }
        sortList(subcategoryList);
        return new Category("Genres", subcategoryList);
    }

    private void sortList(List<Subcategory> list){
        Collections.sort(list, (o1, o2) -> o1.getName().compareTo(o2.getName()));
    }
}

File: app/src/main/java/com/example/moviesearcher/model/repositories/CategoryListRepository.kt
package com.example.moviesearcher.model.repositories

import com.example.moviesearcher.model.data.Category
import com.example.moviesearcher.model.data.Subcategory
import com.example.moviesearcher.model.services.MovieDbApiService
import com.example.moviesearcher.model.services.responses.GenresMapAsyncResponse
import com.example.moviesearcher.model.services.responses.SubcategoryListAsyncResponse
import java.util.*

class CategoryListRepository() {

    private val categoryList: MutableList<Category> = ArrayList()

    interface CategoryListResponse {
        fun onProcessFinished(categoryList: List<Category>?)
    }

    fun fetchData(callback: CategoryListResponse) {
        MovieDbApiService().getGenres(
                GenresMapAsyncResponse { genresMap ->
                    categoryList.add(genreMapToCategory(genresMap!!))

                    MovieDbApiService().getLanguages(SubcategoryListAsyncResponse {
                        sortList(it)
                        categoryList.add(Category("Language", it))
                        callback.onProcessFinished(categoryList)
                    })
                })
    }

    private fun genreMapToCategory(genresMap: HashMap<Int, String>): Category {
        val genreIds: List<Int> = ArrayList(genresMap.keys)
        val genreNames: List<String> = ArrayList(genresMap.values)
        val subcategoryList: MutableList<Subcategory> = ArrayList()
        for (i in genreIds.indices) {
            subcategoryList.add(Subcategory(genreIds[i], genreNames[i]))
        }
        sortList(subcategoryList)
        return Category("Genres", subcategoryList)
    }

    private fun sortList(list: List<Subcategory>) {
        Collections.sort(list) { o1: Subcategory, o2: Subcategory -> o1.name.compareTo(o2.name) }
    }
}